openedx/core/djangoapps/profile_images/images.py

```diff
@@ -0,0 +1,129 @@
+FILE_UPLOAD_BAD_MIMETYPE = _noop('Content-Type header does not match data.')
+
+
+class ImageValidationError(Exception):
+    """
+    Exception to use when the system rejects a user-supplied source image.
+    """
+    @property
+    def user_message(self):
+        """
+        Translate the developer-facing exception message for API clients.
+        """
+        # pylint: disable=translation-of-non-string
+        return _(self.message)
+
+
+def validate_uploaded_image(uploaded_file):
+    """
+    Raises ImageFileRejected if the server should refuse to use this
+    uploaded file as the source image for a user's profile image.  Otherwise,
+    returns nothing.
+    """
+    # validation code by @pmitros,
+    # adapted from https://github.com/pmitros/ProfileXBlock
+    # see also: http://en.wikipedia.org/wiki/Magic_number_%28programming%29
+    image_types = {
+        'jpeg': {
+            'extension': [".jpeg", ".jpg"],
+            'mimetypes': ['image/jpeg', 'image/pjpeg'],
+            'magic': ["ffd8"]
+        },
+        'png': {
+            'extension': [".png"],
+            'mimetypes': ['image/png'],
+            'magic': ["89504e470d0a1a0a"]
+        },
+        'gif': {
+            'extension': [".gif"],
+            'mimetypes': ['image/gif'],
+            'magic': ["474946383961", "474946383761"]
+        }
+    }
+
+    if uploaded_file.size > settings.PROFILE_IMAGE_MAX_BYTES:
+        raise ImageValidationError(FILE_UPLOAD_TOO_LARGE)
+    elif uploaded_file.size < settings.PROFILE_IMAGE_MIN_BYTES:
+        raise ImageValidationError(FILE_UPLOAD_TOO_SMALL)
+
+    # check the file extension looks acceptable
+    filename = unicode(uploaded_file.name).lower()
+    filetype = [ft for ft in image_types if any(filename.endswith(ext) for ext in image_types[ft]['extension'])]
+    if not filetype:
+        raise ImageValidationError(FILE_UPLOAD_BAD_TYPE)
+    filetype = filetype[0]
+
+    # check mimetype matches expected file type
+    if uploaded_file.content_type not in image_types[filetype]['mimetypes']:
+        raise ImageValidationError(FILE_UPLOAD_BAD_MIMETYPE)
+
+    # check magic number matches expected file type
+    headers = image_types[filetype]['magic']
+    if uploaded_file.read(len(headers[0]) / 2).encode('hex') not in headers:
+        raise ImageValidationError(FILE_UPLOAD_BAD_EXT)
+    # avoid unexpected errors from subsequent modules expecting the fp to be at 0
+    uploaded_file.seek(0)
+
+
+def _get_scaled_image_file(image_obj, size):
+    """
+    Given a PIL.Image object, get a resized copy using `size` (square) and
+    return a file-like object containing the data saved as a JPEG.
+
+    Note that the file object returned is a django ContentFile which holds
+    data in memory (not on disk).
+    """
+    if image_obj.mode != "RGB":
+        image_obj = image_obj.convert("RGB")
+    scaled = image_obj.resize((size, size), Image.ANTIALIAS)
```

Comment:
it would make them square (distorting aspect ratio), but we're center-cropping prior to this point when needed.